src/ray/gcs/gcs_server/gcs_job_manager.cc

```diff
@@ -19,56 +19,101 @@
     }
+    RAY_CHECK(state == rpc::JobTableData::RUNNING ||
+              state == rpc::JobTableData::SUBMITTED);
+    job_table_data = iter->second;
+  } else {
+    job_table_data = std::make_shared<JobTableData>();
+  }
+
+  // Reserve these fields that may be filled when submitting.
+  auto is_submitted_from_dashboard = job_table_data->is_submitted_from_dashboard();
+  auto job_payload = job_table_data->job_payload();
+
+  // Just use the job_table_data come from raylet.
+  job_table_data->CopyFrom(request.data());
+  job_table_data->set_state(rpc::JobTableData::RUNNING);
+
+  // Recover reserved fields.
+  job_table_data->set_is_submitted_from_dashboard(is_submitted_from_dashboard);
+  job_table_data->set_job_payload(job_payload);
+
+  auto on_done = [this, job_table_data, driver_pid, reply,
+                  send_reply_callback](const Status &status) {
+    RAY_CHECK_OK(status);
+    auto job_id = JobID::FromBinary(job_table_data->job_id());
+    jobs_.emplace(job_id, job_table_data);
+    RAY_CHECK_OK(gcs_pub_sub_->Publish(JOB_CHANNEL, job_id.Hex(),
+                                       job_table_data->SerializeAsString(), nullptr));
+    RAY_LOG(INFO) << "Finished adding job, job id = " << job_id
+                  << ", driver pid = " << driver_pid;
     GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
   };
-
-  Status status = gcs_table_storage_->JobTable().Put(job_id, request.data(), on_done);
-  if (!status.ok()) {
-    on_done(status);
-  }
+  RAY_CHECK_OK(gcs_table_storage_->JobTable().Put(job_id, *job_table_data, on_done));
 }
 
 void GcsJobManager::HandleMarkJobFinished(const rpc::MarkJobFinishedRequest &request,
                                           rpc::MarkJobFinishedReply *reply,
                                           rpc::SendReplyCallback send_reply_callback) {
   JobID job_id = JobID::FromBinary(request.job_id());
-  RAY_LOG(INFO) << "Marking job state, job id = " << job_id;
-  auto job_table_data =
-      gcs::CreateJobTableData(job_id, /*is_dead*/ true, std::time(nullptr), "", -1);
-  auto on_done = [this, job_id, job_table_data, reply,
-                  send_reply_callback](const Status &status) {
-    if (!status.ok()) {
-      RAY_LOG(ERROR) << "Failed to mark job state, job id = " << job_id;
-    } else {
-      RAY_CHECK_OK(gcs_pub_sub_->Publish(JOB_CHANNEL, job_id.Hex(),
-                                         job_table_data->SerializeAsString(), nullptr));
-      ClearJobInfos(job_id);
-      RAY_LOG(INFO) << "Finished marking job state, job id = " << job_id;
-    }
-    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
-  };
-
-  Status status = gcs_table_storage_->JobTable().Put(job_id, *job_table_data, on_done);
-  if (!status.ok()) {
-    on_done(status);
+  RAY_LOG(INFO) << "Received driver exit notification, job id = " << job_id;
+  auto iter = jobs_.find(job_id);
+  if (iter == jobs_.end()) {
+    RAY_LOG(WARNING) << "Failed to handle the notification of driver exit. job id = "
```

Comment:
"Received a MarkJobFinished request for a job that doesn't exist, job id = xxx."
also, why will this happen?

Reply:
When received a MarkJobFinished  with an incorrect job id. I think it's better not to crash the GCS.